Framework/Algorithms/src/PolarisedSANS/HeliumAnalyserEfficiency.cpp

```diff
@@ -0,0 +1,356 @@
+#include "MantidAPI/HistogramValidator.h"
+#include "MantidAPI/ITableWorkspace.h"
+#include "MantidAPI/WorkspaceUnitValidator.h"
+#include "MantidKernel/BoundedValidator.h"
+#include "MantidKernel/CompositeValidator.h"
+#include "MantidKernel/ListValidator.h"
+
+#include <algorithm>
+#include <boost/algorithm/string.hpp>
+#include <boost/algorithm/string/join.hpp>
+#include <boost/math/distributions/students_t.hpp>
+#include <vector>
+
+namespace Mantid::Algorithms {
+// Register the algorithm into the algorithm factory
+DECLARE_ALGORITHM(HeliumAnalyserEfficiency)
+
+using namespace Kernel;
+using namespace API;
+
+namespace SpinConfigurations {
+static const std::string UpUp = "11";
+static const std::string UpDown = "10";
+static const std::string DownUp = "01";
+static const std::string DownDown = "00";
+} // namespace SpinConfigurations
+
+namespace PropertyNames {
+static const std::string INPUT_WORKSPACE = "InputWorkspace";
+static const std::string OUTPUT_T_WORKSPACE = "OutputTransmissionWorkspace";
+static const std::string P_HE = "HeliumPolarisation";
+static const std::string OUTPUT_T_PARA_WORKSPACE = "OutputTransmissionParaWorkspace";
+static const std::string OUTPUT_T_ANTI_WORKSPACE = "OutputTransmissionAntiWorkspace";
+static const std::string SPIN_CONFIGURATIONS = "SpinConfigurations";
+static const std::string T_E = "TransmissionEmptyCell";
+static const std::string PXD = "GasPressureTimesCellLength";
+static const std::string COVARIANCE = "Covariance";
+static const std::string START_LAMBDA = "StartLambda";
+static const std::string END_LAMBDA = "EndLambda";
+static const std::string STOP_ON_FIT_ERROR = "StopOnFitError";
+} // namespace PropertyNames
+
+/// Empty default constructor algorithm() calls the constructor in the base class
+HeliumAnalyserEfficiency::HeliumAnalyserEfficiency() : Algorithm() {}
+
+void HeliumAnalyserEfficiency::init() {
+  // Declare required input parameters for algorithm and do some validation here
+  auto val = std::make_shared<CompositeValidator>();
+  val->add<WorkspaceUnitValidator>("Wavelength");
+  val->add<HistogramValidator>();
+  declareProperty(std::make_unique<WorkspaceProperty<>>(PropertyNames::INPUT_WORKSPACE, "", Direction::Input, val));
+  declareProperty(std::make_unique<WorkspaceProperty<>>(PropertyNames::OUTPUT_T_WORKSPACE, "T", Direction::Output));
+  declareProperty(std::make_unique<WorkspaceProperty<>>(PropertyNames::P_HE, "p_He", Direction::Output));
+  declareProperty(
+      std::make_unique<WorkspaceProperty<>>(PropertyNames::OUTPUT_T_PARA_WORKSPACE, "T_para", Direction::Output));
+  declareProperty(
+      std::make_unique<WorkspaceProperty<>>(PropertyNames::OUTPUT_T_ANTI_WORKSPACE, "T_anti", Direction::Output));
+
+  std::vector<std::string> initialSpinConfig{
+      {SpinConfigurations::UpUp, SpinConfigurations::UpDown, SpinConfigurations::DownUp, SpinConfigurations::DownDown}};
+  std::sort(initialSpinConfig.begin(), initialSpinConfig.end());
```

Comment:
From the docs for `std::next_permutation` it looks like it returns permutations in lexicographical order, so if the initial vector isn't sorted then `next_permutation` will stop early, without giving all the possible permutations. But I'm thinking that it might be better to just use a custom validator that lets the user type in a string, instead of the combo box setup I used here. Although I'll probably try and manoeuvre in a use of this permutation stuff, seems like a waste otherwise.